bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceRegistry.java

```diff
@@ -1072,8 +1073,16 @@ void removeServiceRegistration(BundleContextImpl context, ServiceRegistrationImp
 	private List<ServiceRegistrationImpl<?>> lookupServiceRegistrations(String clazz, Filter filter) {
 		List<ServiceRegistrationImpl<?>> result;
 		synchronized (this) {
-			if (clazz == null) { /* all services */
-				result = allPublishedServices;
+			if (clazz == null) {
+				if (filter instanceof FilterImpl) {
+					// check if we can determine the clazz from the filter
+					String filterObjectClazz = ((FilterImpl) filter).getRequiredObjectClass();
```

Comment:
This is not correct. `getRequiredObjectClass()` only returns the first `objectClass` value observed in the filter expression. Which means that this change will return the wrong result for an expression like:

`(|(objectClass=org.acme.BrickService)(markerProperty=*))`

Due to the complexity of filter expressions it is very hard to correctly "collapse" a filter expression onto a single objectClass value. So there is no (current) way to optimized this code path.

Reply:
I'll have to add a testcase for that then because I thought that would work correctly.

Reply:
For example Device Manager does `"(|(objectClass=org.osgi.service.device.Device)(DEVICE_CATEGORY=*))"` to locate device services.

Reply:
I've added a testcase that shows it working.  Let me know if you think other filter expression examples should be tested.